CTSKF-710 feedback model refactor

AlexRyanShep opened the pull request:
#### What
Refactor the Feedback model/controller, ZendeskSender and SurveyMonkey sender to remove responsibility for error handling and decision of what sending service to use from the model. 

#### Ticket

[CTSKF-710](https://dsdmoj.atlassian.net/browse/CTSKF-710)

#### Why

- Make the Feedback model conform to SOLID principles and follow the model > services design pattern. 
- Simplify the process for adding new sending services and/or switching between zendesk and surveymonkey for feedback. 

#### How

Model, Controller and Service files refactored to move logic around and use dependency inversion to provide the sender class for each piece of feedback.



[CTSKF-710]: https://dsdmoj.atlassian.net/browse/CTSKF-710?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

Katy600 reviewed (commented):
Hi @AlexRyanShep, I enjoyed reading your code as it felt really clear. Obviously take all my comments lightly as I am just trying to improve my code reviews as well as how I read code. And please correct any misconception I have.

AlexRyanShep commented on spec/requests/feedback_spec.rb:
```diff
@@ -45,6 +45,32 @@
+
+        it 'Uses the Zendesk sender' do
+          expect(Feedback).to have_received(:new).with(anything, ZendeskSender)
+        end
```
I don't think that creating two different users would work here, the toggle between Zendesk and SurveyMonkey for Feedback is set at the application level in the environment variables in `.k8s/live/production/app-config.yaml`, not at the user level. That's why these tests have the in the before block, `allow(Settings).to receive(:zendesk_feedback_enabled?).and_return(true)`

Katy600 replied:
Oh I see, thanks Alex, that makes sense.

AlexRyanShep commented on spec/requests/feedback_spec.rb:
```diff
@@ -45,6 +45,32 @@
       }
     end
 
+    context 'when feedback_controller is determining the sender class to use' do
```
Makes sense to me 👍  I'll do some re-writing to remove the unnecessary contexts

Katy600 replied:
Oh great!

Katy600 commented on spec/services/zendesk_sender_spec.rb:
```diff
@@ -33,14 +34,17 @@
-    context 'with a reporter email' do
-      let(:ticket_payload) do
+    context 'with a valid ticket and reporter email' do
+      let(:bug_report_ticket_payload) do
```
Is this the same bug_report_ticket_payload as the one above? Are you able to reuse it somehow?

AlexRyanShep replied:
This is code that was already in this test file that I've just tweaked. The difference between them is that the first instance of the payload has `nil` for the email address, while this one has is populated. I assume this is to make sure that it can run with and without an email without raising an error.

@jrmhaig git blame shows you as the original author of most of this file, am I understanding the purpose of that right?

AlexRyanShep commented:
> Hi @AlexRyanShep, I enjoyed reading your code as it felt really clear. Obviously take all my comments lightly as I am just trying to improve my code reviews as well as how I read code. And please correct any misconception I have.

Thanks @Katy600! Good spot with the Hash thing, I got too caught up in the original way that I solved that I think to recognise there was a simpler one!

I've just pushed a new commit that addresses the points you made :-)

Katy600 commented on app/services/zendesk_sender.rb:
```diff
@@ -1,17 +1,28 @@
 
   def initialize(ticket_payload)
     @ticket_payload = ticket_payload
+    @return_hash = {}
```
Does '@return_hash' need to be an instance variable? Could it just be a local variable inside the call method? And then it could just be passed to catch_error. I am asking this to check my understanding of when to use an instance variable as I am not entirely sure myself.

AlexRyanShep replied:
My logic was just that having it be an instance variable is a more elegant solution than having to have it be passed to the catch_error, and then catch_error would need to return it back so that call could return it.

That said, looking at it again has made me think of what I think is a nicer way of doing it. I've amended it onto the most recent commit, what do you think?

Katy600 replied:
Oh I like that @AlexRyanShep. I would be interested in other peoples thoughts but it feels more readable having it as a whole hash inside the method - where you can see all the values at once. And more encapsulated perhaps!?!?

AlexRyanShep commented on app/models/feedback.rb:
```diff
@@ -7,18 +7,23 @@ class Feedback
   validates :event, :outcome, presence: true, if: :bug_report?
 
-  def initialize(attributes = {}, sender = nil)
+  def initialize(sender = nil, attributes = {})
```
New commit now submits :sender as a part of the existing attributes hash rather than on its own. Validation only required slight tweaking, as a non-existent :sender attribute is nil, which can still be checked against the list of acceptable senders

jrmhaig reviewed (commented):
Thanks, @AlexRyanShep. This looks great. I have made a few comments and I'd be happy to discuss them if you want.

Ideally it would be good if the `type` parameter were not required by feedback model at all. This would require some more refactoring so it is not necessary for this PR, unless you think it can be done simply. Some things to think about regarding this;

* The list of fields stored in the `FEEDBACK_TYPES` array are proper to the sender so instead of `FEEDBACK_TYPES[type.to_sym]` you could have `@sender.feedback_fields`.
* Following the above point, if the sender classes have the list of fields that are applicable then, in the case of Zendesk, the `subject`, `description` and `reporter_email` methods can also go in the sender class.

AlexRyanShep commented on app/services/survey_monkey_sender.rb:
```diff
@@ -9,7 +9,18 @@ def initialize(feedback)
 
   def call
-    response.submit
+    return_hash = {}
+    resp = response.submit
+
+    return_hash[:success] = resp[:success]
+
+    return_hash[:response_message] = if resp[:success]
+                                       'Feedback submitted'
+                                     else
+                                       "Unable to submit feedback [#{resp[:error_code]}]"
+                                     end
+
+    return_hash
```
Looks good to me, thanks!

AlexRyanShep commented on app/services/zendesk_sender.rb:
```diff
@@ -1,17 +1,24 @@
-    def send!(ticket_payload)
-      ZendeskSender.new(ticket_payload).send!
-    end
+  def self.call(ticket_payload)
+    new(ticket_payload).call
```
This is handy, thanks!

AlexRyanShep commented on app/models/feedback.rb:
```diff
@@ -18,6 +23,7 @@ def initialize(attributes = {})
     attributes.each do |key, value|
       instance_variable_set(:"@#{key}", value)
     end
+    @sender = SENDING_SERVICES.include?(@sender) ? @sender : nil
```
Implemented this recommendation during our call

AlexRyanShep commented:
@jrmhaig my latest commit implements most of the recommendations that you made, including moving the #subject, #description etc methods into the ZendeskSender class, I managed to figure out what was stopping them working shortly after you had to leave.

The only remaining thing I've not done is removal of the Type parameter from the Feedback model. I've looked more into how this is used and they're very closely linked. Type is not only used to determine which sender is used, but it is also used to determine which of the feedback templates are rendered when the page is loaded (within whatever haml file generates the link,  `new_feedback_path(type: 'bug_report')` for example).

The only way I can think of right now to decouple Feedback from the type parameter would be to create a new BugReport class that inherits from Feedback, but then that makes me start thinking about problems with duplication of code  in the controller for that new class and so on. Let's discuss tomorrow, but that last little bit there is feeling like more of a rabbit hole than it might be worth to me .

jrmhaig reviewed (approved):
If you haven't done already it would be good to put this on staging and try submitting some bug reports and feedback. Otherwise it looks good to go.

AlexRyanShep commented:
> If you haven't done already it would be good to put this on staging and try submitting some bug reports and feedback. Otherwise it looks good to go.

Thanks Joe! I've done a bit as I've gone, but I'll push it into those environments and do a last round of testing before I merge

AlexRyanShep commented:
I've done some testing, staging is sending feedback and bug reports through to SurveyMonkey and Zendesk fine. Merging into master now
